### DISCUSSION TITLE
DependencyMatcher Optional Tokens

### DISCUSSION THREAD

**Discussion by Ptark** [explosion/spaCy#11614] [Help: Coding & Implementations] [answered]
Hi, I have a usecase where I want to match German "Futur I" but not "Futur II". I use the following pattern:
```
FUTURE_PATTERNS = [[
    {
        "RIGHT_ID": "anchor_future_one_werden",
        "RIGHT_ATTRS": {"LEMMA": "werden", "DEP": "ROOT"}
    },
    {
        "LEFT_ID": "anchor_future_one_werden",
        "REL_OP": ">",
        "RIGHT_ID": "infinitive",
        "RIGHT_ATTRS": {
            "TAG": {"IN": INFINITIVE_TAGS},
            "DEP": "oc"
        }
    },
    {
        "LEFT_ID": "infinitive",
        "REL_OP": ">",
        "RIGHT_ID": "not_participle",
        "RIGHT_ATTRS": {
            "TAG": {"NOT_IN": PARTICIPLE_TAGS},
            "DEP": {"IN": ["oc, "pd"]}
        }
    }
]]
```
This matches Futur I if the infinitive still has some child token that is not a participle. But it doesn't match if there are no child tokens. Can I somehow make the 3rd token optional? "OP": "?" does not work.

This is related to the discussions: 
- https://github.com/explosion/spaCy/discussions/6783
- https://github.com/explosion/spaCy/discussions/8045
Basically I would like an optional Leaf Node. Has this feature been implemented?

**Comment by polm** [comment, marked as answer]
As mentioned in the threads you link to / posted in, the DependencyMatcher can only use patterns such that each node matches exactly one token. `OP: ?` could match 0 tokens and so it does not work. 

If you want to make the third token optional, what you can do is simply add two patterns with the same label. The first pattern would have just two tokens, and the second pattern would have all three tokens.

**Reply by Ptark** [reply]
Thank you for your answer and sorry for double posting. Your suggestion of using a 2 token and a 3 token pattern doesn't exactly cover my usecase, because I want to exclude tokens in the 3 token pattern which would then again be included in the 2 token pattern.

Your suggestion gave me the idea however, to instead search for unwanted tokens in the 3 token pattern and then exclude matches where the first 2 tokens in both patterns align.

**Reply by polm** [reply]
Yes, you'd need to filter the overlaps like that. Glad it works for you!
